# **CLASSIFICATION**





In [2]:
text = """ code geass is one of those series that everybody recomends and once you watch it you know why.

Mechas, war, romance, comedy, rivality, unexpected twsits, explotions, mindgames and more mechas! what else do you want!!! xD it totally deserver its place as one of the best animes ever made.

the only aspect i see as a flaw is the art with a 9/10 the people seem a little bit deformed or too skinny but once you get use to it is great!.The user wants freedom.

and it has by far the best story i have ever seen you just get hooked to it from the second episode and cant stop until finishing it.

almost perfect serie, recomended to anyone who is a human beeing (nevermind, my dog liked it) """

In [5]:
!pip install pyarrow==14.0.1 requests==2.31.0

In [4]:
!pip install transformers
!pip install datasets==2.1.0
!pip install sentencepiece

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 325.4/325.4 kB 8.9 MB/s eta 0:00:00
  Attempting uninstall: datasets
    Found existing installation: datasets 2.20.0
    Uninstalling datasets-2.20.0:
      Successfully uninstalled datasets-2.20.0


# **Transformers:**

- The Transformers library, developed by Hugging Face, provides state-of-the-art natural language processing (NLP) models, including pre-trained models like BERT, GPT, RoBERTa, and more.<br><br>
- These models are built using transformer architectures, which have revolutionized NLP by achieving superior performance on a wide range of tasks such as text classification, question answering, text generation, and language translation.<br><br>
- The Transformers library allows users to easily load, fine-tune, and use these pre-trained models for various NLP tasks, enabling researchers and developers to leverage the latest advancements in deep learning for their projects.<br><br>
- It also provides interfaces for working with these models in popular deep learning frameworks like TensorFlow and PyTorch.<br><br>
# **Datasets:**

- The Datasets library, also developed by Hugging Face, provides a collection of high-quality datasets for training and evaluating NLP models.<br><br>
- These datasets cover a wide range of tasks and domains, including text classification, named entity recognition, machine translation, and more.<br><br>
- The Datasets library offers a unified interface for accessing and manipulating these datasets, making it easy for researchers and practitioners to experiment with different datasets and benchmark their models against established baselines.<br><br>
- It provides functionality for downloading datasets, splitting them into train/validation/test sets, preprocessing text data, and loading datasets in a format compatible with popular deep learning frameworks.<br><br>

# **SentencePiece:**

- SentencePiece is a library developed by Google for tokenization and text normalization tasks, particularly in the context of neural network-based models.<br><br>
- It offers methods for segmenting text into subword units, which can be useful for handling out-of-vocabulary words, reducing vocabulary size, and improving model generalization.<br><br>
- SentencePiece uses the unigram language model to learn a tokenization scheme from the input text data, allowing it to effectively handle various languages and writing systems.<br><br>



      Training the SentencePiece Model:

      Imagine we have a big collection of text, like books or articles.
      We use this text to teach the SentencePiece model how words are formed.
      The model learns patterns in the text and decides how to split words into smaller parts called subwords.
      Tokenization of New Text:

      After training, we have a model that knows how to break down words into smaller pieces.
      When we give it a new sentence, like "The cat is sitting on the mat," it splits it into these smaller pieces based on what it learned during training.
      For example, it might split "sitting" into "sit" and "ting" because it didn't see "sitting" before, but it knows "sit" and "ting."
      Example Tokenization:

      Let's say our SentencePiece model knows the subwords "the," "cat," "is," "sit," "ting," "on," "mat," and ".".
      When it tokenizes "The cat is sitting on the mat," it might output: ["the", "cat", "is", "sit", "ting", "on", "the", "mat", "."].
      Notice how it split "sitting" into "sit" and "ting" because it didn't know "sitting" as a whole.
      Handling New Words:

      If the model encounters a word it didn't see during training, it tries to represent it using the subwords it knows.
      For example, if it sees a new word like "running," it might split it into "run" and "ning."
      In short, SentencePiece learns from a lot of text to understand how words are built and then uses that knowledge to break down new text into smaller pieces. This helps it handle new words and make text easier for computers to understand.
- This library is commonly used in conjunction with other NLP libraries and frameworks, such as TensorFlow and PyTorch, to preprocess text data before training or inference with deep learning models.

In [3]:
import datasets
import huggingface_hub
import matplotlib.pyplot as plt
import transformers

- huggingface_hub democratizes access to cutting-edge NLP models and datasets by providing a centralized repository where users can easily discover and download resources for their projects.



- By providing a centralized hub for hosting models and datasets, huggingface_hub streamlines the process of sharing and accessing resources, thereby promoting collaboration, reproducibility, and innovation in the field of NLP.

In [4]:
# Text Classification
from transformers import pipeline
classifier = pipeline('text-classification')

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that aut

# **Transformers Pipeline**

- The core architecture behind the text classification pipeline in the Hugging Face Transformers library typically involves a pre-trained transformer model fine-tuned on a large corpus of text data for the specific task of text classification. Here's how it works, illustrated with an example:<br><br>

**Loading the Model:**

- When you create a text classification pipeline using pipeline('text-classification'), the Hugging Face library automatically downloads the pre-trained model associated with text classification tasks. This model is typically a transformer-based architecture like BERT, RoBERTa, or DistilBERT that has been fine-tuned on a dataset for text classification.<br><br>


**Tokenization:**

- The input text provided to the pipeline is tokenized using the same tokenizer that was used during the pre-training of the model. This tokenizer converts the input text into a sequence of tokens that can be processed by the transformer model.<br><br>


**Encoding:**

- The tokenized input text is then encoded into numerical representations suitable for input to the transformer model. This encoding typically involves mapping each token to its corresponding index in the model's vocabulary and converting the input sequence into numerical vectors.<br><br>


**Model Inference:**

- The encoded input sequence is fed into the pre-trained transformer model. The model processes the input sequence through multiple layers of self-attention mechanisms and feed-forward neural networks to extract contextual representations of the text.<br><br>


**Classification Head:**

- After processing the input text through the transformer layers, a classification head is attached to the model. This classification head consists of one or more dense layers that take the final contextual representation of the input text as input and produce predictions for the classification task.<br><br>
- For example, in binary classification tasks, the classification head may consist of a single neuron with a sigmoid activation function that outputs a probability score indicating the likelihood of the input text belonging to a particular class.<br><br>
- In multi-class classification tasks, the classification head may consist of multiple neurons corresponding to each class, with a softmax activation function applied to produce a probability distribution over the classes.
Prediction:

Finally, the model outputs the predicted class label or probability distribution over the classes for the input text. This prediction is returned as the output of the text classification pipeline.

**config.json:**

- This file contains the configuration parameters of the model, including details such as the model architecture, hyperparameters, tokenizer settings, and any other model-specific configurations. It provides essential information needed to instantiate and use the model in your code.

        {
          "attention_probs_dropout_prob": 0.1,    This parameter determines the dropout probability for attention weights. Dropout is a regularization technique used during training to prevent overfitting by randomly setting some activations to zero.
          "hidden_dropout_prob": 0.1,   Similar to attention_probs_dropout_prob, this parameter specifies the dropout probability for the hidden layers. It controls the dropout applied to the outputs of each layer in the model.
          "hidden_size": 768,   This parameter defines the dimensionality of the hidden layers in the transformer model. It represents the number of neurons in each layer's feedforward network.
          "num_attention_heads": 12,
          "num_hidden_layers": 12,   The total number of layers in the transformer model, including both encoder and decoder layers. Each layer consists of sublayers like self-attention, feedforward, and layer normalization.
          "type_vocab_size": 2,
          "vocab_size": 30522,
          ...
        }


**model.safetensors**


- safetensors is a safe and fast file format for storing and loading tensors. Typically, PyTorch model weights are saved or pickled into a .bin file with Python’s pickle utility. However, pickle is not secure and pickled files may contain malicious code that can be executed. safetensors is a secure alternative to pickle, making it ideal for sharing model weights.<br><br>


          # Shopping cart data
          shopping_cart = {
              "items": [
                  {"name": "Laptop", "quantity": 1, "price": 999.99},
                  {"name": "Headphones", "quantity": 2, "price": 49.99},
                  {"name": "Mouse", "quantity": 1, "price": 19.99}
              ],
              "total": 1119.96
          }


          import json

          # Serialize the shopping cart to JSON
          serialized_cart = json.dumps(shopping_cart)

          # Save the serialized data to a file
          with open("shopping_cart.json", "w") as file:
              file.write(serialized_cart)



                        # Read the serialized data from the file
          with open("shopping_cart.json", "r") as file:
              serialized_cart = file.read()

          # Deserialize the JSON data into a Python dictionary
          restored_cart = json.loads(serialized_cart)

          print(restored_cart)



**pytorch_model.bin:**

- This file contains the actual weights and parameters of the BERT model, learned during pre-training or fine-tuning.
Example content: Serialized tensor data representing the weights of each layer, attention matrices, and other parameters of the model.

- Serialization: Serialization is the process of converting complex data structures, such as tensors, into a format that can be easily stored, transmitted, or reconstructed later. Serialization typically involves converting the data into a byte stream or a string format that preserves the structure and content of the original data.<br><br>

        import pickle

        # Example data to serialize
        data = {'name': 'John', 'age': 30, 'city': 'New York'}

        # Serialize data using pickle and save to a file
        with open('data.pkl', 'wb') as file:
            pickle.dump(data, file)

        # Deserialize data from the file
        with open('data.pkl', 'rb') as file:
            loaded_data = pickle.load(file)

        print(loaded_data)



**tokenizer_config.json:**

- This file contains the configuration parameters of the BERT tokenizer, such as the tokenizer type, special tokens, vocabulary size, etc.
Example content:


          {
            "max_len": 512,    This indicates the maximum sequence length that the tokenizer will process. Sequences longer than this length will be truncated, and sequences shorter than this length will be padded to reach this length.
            "model_type": "bert",
            "pad_token_id": 0,    This is the token ID used for padding sequences. When sequences are shorter than the maximum length, they are padded with this token ID to match the maximum length.
            "vocab_size": 30522,
            ...
          }


In [5]:
classifier(text)

[{'label': 'POSITIVE', 'score': 0.9967235922813416}]

# **NER - Named Entity Recoginition**

In [6]:
import pandas as pd

In [7]:
text1 = "Narendra Modi is prime minister of India"

In [8]:
ner = pipeline('ner', aggregation_strategy='simple')
out = ner(text1)
print(pd.DataFrame(out))

No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision f2482bf (https://huggingface.co/dbmdz/bert-large-cased-finetuned-conll03-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


  entity_group     score           word  start  end
0          PER  0.998214  Narendra Modi      0   13
1          LOC  0.997805          India     35   40


In the Hugging Face Transformers library, the aggregation_strategy parameter in the Named Entity Recognition (NER) pipeline specifies how multiple predictions for the same entity across different tokens are aggregated into a single entity span.




Let's illustrate the   **aggregation_strategy='simple'**    parameter in the Named Entity Recognition (NER) pipeline with an example:<br><br>










Suppose we have the following input text:


      "The company Google is headquartered in Mountain View, California."


And let's assume the NER model predicts the following entity labels for each token:

      Token:         The   company  Google   is    headquartered   in    Mountain    View       ,        California   .
      Predicted NER: O     O        B-ORG    O     O              O     B-LOC       I-LOC      O        B-LOC        .


The output of the pipeline would be a list of dictionaries, with each dictionary representing an entity found in the input text. For example:


        [
          {'entity': 'ORG', 'text': 'Google', 'start': 12, 'end': 18},
          {'entity': 'LOC', 'text': 'Mountain View', 'start': 34, 'end': 47},
          {'entity': 'LOC', 'text': 'California', 'start': 53, 'end': 63}
        ]




## **Question Answering**

In [9]:
reader = pipeline('question-answering')
question = 'what does user want ?'
outputs = reader(question=question,context = text)
pd.DataFrame([outputs])

No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 626af31 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


score  start  end   answer
0  0.932237    450  457  freedom

## **Summarization**

In [10]:
summarizer = pipeline('summarization')
outputs = summarizer( text, clean_up_tokenization_spaces=True, max_length=90)
print(outputs[0]['summary_text'])

No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


 The only aspect i see as a flaw is the art with a 9/10 the people seem a little bit deformed or too skinny but once you get use to it is great. The almost perfect serie, recomended to anyone who is a human beeing (nevermind, my dog liked it)


**clean_up_tokenization_spaces=True:**<br><br>

- When clean_up_tokenization_spaces is set to True, it indicates that the pipeline should remove any extra spaces resulting from tokenization before generating the summary text.<br><br>
- Tokenization of text involves splitting it into individual tokens (words or subwords). Sometimes, tokenization can introduce extra spaces, especially around punctuation marks or special characters.<br><br>
- Setting clean_up_tokenization_spaces=True ensures that these extra spaces are removed from the input text before it is summarized, resulting in cleaner and more readable summaries.<br><br>


**max_length=90:**<br><br>

- The max_length parameter specifies the maximum length (in tokens) of the generated summary text.<br><br>
- When summarizing text, it's often desirable to limit the length of the summary to make it concise and focused.<br><br>
- By setting max_length=90, you're instructing the summarization pipeline to generate a summary with a maximum length of 90 tokens. If the summary exceeds this length, it will be truncated to meet the specified limit.<br><br>
- Limiting the length of the summary can help control the amount of information presented to the reader and ensure that the summary remains informative and easy to comprehend.

## **Translation**

In [11]:
translate = pipeline('translation_en_to_de', model='Helsinki-NLP/opus-mt-en-de')
output = translate(text,clean_up_tokenization_spaces=True,min_length=120)
print(output[0]['translation_text'])

/usr/local/lib/python3.10/dist-packages/transformers/models/marian/tokenization_marian.py:175: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")


Code Geass ist eine jener Serien, die jeder recomends und sobald Sie es sehen Sie wissen, warum. Mechas, Krieg, Romantik, Komödie, Rivalität, unerwartete Twsits, Explotionen, Mindgames und mehr mechas! was sonst wollen Sie! xD es völlig verdient seinen Platz als einer der besten Animes jemals gemacht. der einzige Aspekt, den ich als Fehler sehe, ist die Kunst mit einem 9/10 die Menschen scheinen ein wenig verformt oder zu dünn, aber wenn Sie es nutzen ist toll!.Der Benutzer will Freiheit. und es hat bei weitem die beste Geschichte, die ich je gesehen habe, Sie nur an sie von der zweiten Folge und Cant Stop bis zum Ende. fast perfekt serie, recomended zu jedem, der ein Mensch ist Bienen (Nevermind, mein Hund mochte es)


## **TEXT CLASSIFICATION WITH HUGGINGFACE**

In [12]:
!pip install --upgrade pip
!pip install transformers
!pip install datasets
!pip install sentencepiece

# Datasets

In [13]:
# Emotion detector
from datasets import list_datasets
all_datasets = list_datasets()
print(f"There are total of {len(all_datasets)} in the hub")
print(f"Some example datasets are {all_datasets[:5]}")

<ipython-input-13-d8db1cb5906a>:3: FutureWarning: list_datasets is deprecated and will be removed in the next major version of datasets. Use 'huggingface_hub.list_datasets' instead.
  all_datasets = list_datasets()


There are total of 168484 in the hub
Some example datasets are ['amirveyseh/acronym_identification', 'ade-benchmark-corpus/ade_corpus_v2', 'UCLNLP/adversarial_qa', 'Yale-LILY/aeslc', 'nwu-ctext/afrikaans_ner_corpus']


In [14]:
from datasets import load_dataset
# Load the 'go_emotions' dataset
emotion_dataset = load_dataset('go_emotions')
print(emotion_dataset)

Generating train split:   0%|          | 0/43410 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/5426 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/5427 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['text', 'labels', 'id'],
        num_rows: 43410
    })
    validation: Dataset({
        features: ['text', 'labels', 'id'],
        num_rows: 5426
    })
    test: Dataset({
        features: ['text', 'labels', 'id'],
        num_rows: 5427
    })
})


In [15]:
train_dataset = emotion_dataset['train']
train_dataset

Dataset({
    features: ['text', 'labels', 'id'],
    num_rows: 43410
})

In [16]:
train_dataset[0]

{'text': "My favourite food is anything I didn't have to cook myself.",
 'labels': [27],
 'id': 'eebbqej'}

In [22]:
train_dataset[:5]

{'text': ["My favourite food is anything I didn't have to cook myself.",
  'Now if he does off himself, everyone will think hes having a laugh screwing with people instead of actually dead',
  'WHY THE FUCK IS BAYLESS ISOING',
  'To make her feel threatened',
  'Dirty Southern Wankers'],
 'labels': [[27], [27], [2], [14], [3]],
 'id': ['eebbqej', 'ed00q6i', 'eezlygj', 'ed7ypvh', 'ed0bdzj']}

In [17]:
train_dataset.column_names

['text', 'labels', 'id']

In [18]:
train_dataset.features

{'text': Value(dtype='string', id=None),
 'labels': Sequence(feature=ClassLabel(names=['admiration', 'amusement', 'anger', 'annoyance', 'approval', 'caring', 'confusion', 'curiosity', 'desire', 'disappointment', 'disapproval', 'disgust', 'embarrassment', 'excitement', 'fear', 'gratitude', 'grief', 'joy', 'love', 'nervousness', 'optimism', 'pride', 'realization', 'relief', 'remorse', 'sadness', 'surprise', 'neutral'], id=None), length=-1, id=None),
 'id': Value(dtype='string', id=None)}

In [19]:
print(train_dataset.column_names)

['text', 'labels', 'id']


In [20]:
# Assuming 'labels' column contains lists and you want unique labels across all lists
unique_labels = set([label for sublist in train_dataset['labels'] for label in sublist])
print(unique_labels)

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27}


In [21]:
!wget https://raw.githubusercontent.com/nachikethmurthy/Source-Code-Dataset-for-Machine-Learning-using-Python/main/Data/sentiment_train

--2024-06-25 06:55:57--  https://raw.githubusercontent.com/nachikethmurthy/Source-Code-Dataset-for-Machine-Learning-using-Python/main/Data/sentiment_train
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 447555 (437K) [text/plain]
Saving to: ‘sentiment_train.1’

sentiment_train.1   100%[===================>] 437.07K  --.-KB/s    in 0.01s   

2024-06-25 06:55:57 (39.5 MB/s) - ‘sentiment_train.1’ saved [447555/447555]



In [1]:
!pip install --upgrade datasets

In [35]:
!pip install --upgrade pandas


In [22]:
import pandas as pd
# load the dataset
sentiment_train = load_dataset('csv',data_files = "sentiment_train",sep='\t')
sentiment_train

Generating train split: 0 examples [00:00, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['sentiment', 'text'],
        num_rows: 6918
    })
})

In [23]:
emotion_dataset

DatasetDict({
    train: Dataset({
        features: ['text', 'labels', 'id'],
        num_rows: 43410
    })
    validation: Dataset({
        features: ['text', 'labels', 'id'],
        num_rows: 5426
    })
    test: Dataset({
        features: ['text', 'labels', 'id'],
        num_rows: 5427
    })
})

Convert the Hugging Face Dataset object (emotion_dataset) into a pandas DataFrame (df)

In [24]:
import pandas as pd
emotion_dataset.set_format(type='pandas')
df = emotion_dataset['train'][:]
df.head()

text labels       id
0  My favourite food is anything I didn't have to...   [27]  eebbqej
1  Now if he does off himself, everyone will thin...   [27]  ed00q6i
2                     WHY THE FUCK IS BAYLESS ISOING    [2]  eezlygj
3                        To make her feel threatened   [14]  ed7ypvh
4                             Dirty Southern Wankers    [3]  ed0bdzj

In [25]:
emotion_dataset['train']

Dataset({
    features: ['text', 'labels', 'id'],
    num_rows: 43410
})

In [26]:
emotion_dataset['train'][:]

text labels       id
0      My favourite food is anything I didn't have to...   [27]  eebbqej
1      Now if he does off himself, everyone will thin...   [27]  ed00q6i
2                         WHY THE FUCK IS BAYLESS ISOING    [2]  eezlygj
3                            To make her feel threatened   [14]  ed7ypvh
4                                 Dirty Southern Wankers    [3]  ed0bdzj
...                                                  ...    ...      ...
43405  Added you mate well I’ve just got the bow and ...   [18]  edsb738
43406  Always thought that was funny but is it a refe...    [6]  ee7fdou
43407  What are you talking about? Anything bad that ...    [3]  efgbhks
43408            More like a baptism, with sexy results!   [13]  ed1naf8
43409                                    Enjoy the ride!   [17]  eecwmbq

[43410 rows x 3 columns]

In [30]:
print(emotion_dataset['train'].features)

{'text': Value(dtype='string', id=None), 'labels': Sequence(feature=ClassLabel(names=['admiration', 'amusement', 'anger', 'annoyance', 'approval', 'caring', 'confusion', 'curiosity', 'desire', 'disappointment', 'disapproval', 'disgust', 'embarrassment', 'excitement', 'fear', 'gratitude', 'grief', 'joy', 'love', 'nervousness', 'optimism', 'pride', 'realization', 'relief', 'remorse', 'sadness', 'surprise', 'neutral'], id=None), length=-1, id=None), 'id': Value(dtype='string', id=None)}


In [31]:
print(emotion_dataset['train'][:5])

                                                text labels       id
0  My favourite food is anything I didn't have to...   [27]  eebbqej
1  Now if he does off himself, everyone will thin...   [27]  ed00q6i
2                     WHY THE FUCK IS BAYLESS ISOING    [2]  eezlygj
3                        To make her feel threatened   [14]  ed7ypvh
4                             Dirty Southern Wankers    [3]  ed0bdzj


In [32]:
# Get the first label of the first example
first_label = emotion_dataset['train'][0]['labels'][0]
label_name = emotion_dataset['train'].features['labels'].feature.int2str(first_label)
print(label_name)

['neutral']


**The int2str()**<br><br>


 method is used to convert integer indices back to string labels. It's particularly useful when dealing with classification tasks where labels are represented as integers but you want to convert them back to their original string representations.

In [33]:
# Get the second label (if it exists) of the first example
example = emotion_dataset['train'][0]
if len(example['labels']) > 1:
    second_label = example['labels'][1]
    label_name = emotion_dataset['train'].features['labels'].feature.int2str(second_label)
    print(label_name)
else:
    print("This example doesn't have a second label")

This example doesn't have a second label


In [35]:
def label_to_str(row):
  return emotion_dataset['train'].features['labels'].feature.int2str(row) # Change 'label' to 'labels'

df['labels_name'] = df['labels'].apply(label_to_str) # Change 'label' to 'labels'
df.head()

text labels       id  \
0  My favourite food is anything I didn't have to...   [27]  eebbqej   
1  Now if he does off himself, everyone will thin...   [27]  ed00q6i   
2                     WHY THE FUCK IS BAYLESS ISOING    [2]  eezlygj   
3                        To make her feel threatened   [14]  ed7ypvh   
4                             Dirty Southern Wankers    [3]  ed0bdzj   

   labels_name  
0    [neutral]  
1    [neutral]  
2      [anger]  
3       [fear]  
4  [annoyance]

In [36]:
emotion_dataset.reset_format()   #Hugging Face datasets library, such as the DatasetDict

In [ ]:
# Tokenization - Character tokenization
# Tokenization - Word tokenization
# Goal --> Find a way to effienctly perform the tokenization - Subword Tokenization

In [37]:
from transformers import AutoTokenizer
model = 'distilbert-base-uncased'
tokenizer = AutoTokenizer.from_pretrained(model)

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

In [38]:
text = "My favorite food is Masala Dosa"
encoded_text = tokenizer(text)
print(encoded_text)

{'input_ids': [101, 2026, 5440, 2833, 2003, 16137, 7911, 9998, 2050, 102], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}


      input_ids: This is a list of token IDs representing the input text after tokenization. Each token in the input text is mapped to a corresponding ID from the model's vocabulary. Here's what each ID represents:

      101: This is the special token [CLS], which stands for "classification" and is added at the beginning of the input sequence.
      2026: The token ID for the word "My".
      5440: The token ID for the word "favorite".
      2833: The token ID for the word "food".
      2003: The token ID for the word "is".
      16137: The token ID for the subword "mas".
      7911: The token ID for the subword "##ala".
      9998: The token ID for the subword "dos".
      2050: The token ID for the subword '##a'
      102: This is another instance of the special token [SEP], indicating the end of the sequence.
      attention_mask: This is a list indicating which tokens in the input sequence should be attended to (have their attention scores calculated) and which ones should be ignored. In this case, all tokens have an attention mask value of 1, indicating that they should all be attended to.

**[CLS] stands for "classification"**


- The special token [CLS] stands for "classification" and is added at the beginning of the input sequence specifically for tasks that involve classification or sequence labeling, such as sentiment analysis, text classification, or named entity recognition. Its main purpose is to provide a representation of the entire input sequence that can be used for classification tasks.

- When training a model for classification tasks using techniques like BERT, the model typically receives a single vector representation of the entire input sequence. The [CLS] token's position is chosen because it always appears in the same position in every input sequence, making it a consistent representation to use for classification purposes<br><br>


**[SEP] Token (ID 102):**

- The [SEP] token is indeed a separator token used to mark the end of a sequence or to separate segments within a sequence.
- It's added at the end of each input sequence to indicate the end of the sequence.
- In tasks like sentence classification or question answering, where there's only one input sequence, a single [SEP] token is added at the end to denote the end of that sequence.





In [39]:
type(encoded_text)

transformers.tokenization_utils_base.BatchEncoding

**transformers:**

This is the namespace or package where the class tokenization_utils_base is defined. In the context of Hugging Face's Transformers library, this is where the tokenization-related utilities are implemented.<br><br>

**tokenization_utils_base:**


This is a module within the Transformers library that contains base classes and utilities for tokenization. It provides the foundation for various tokenization-related functionalities.<br><br>

**BatchEncoding:**

This is the specific class or type of object returned by the tokenizer when encoding a batch of input sequences. It represents a batch of encoded sequences and typically contains attributes such as input_ids, attention_mask, and other relevant information about the encoded batch.

In [40]:
tokens = tokenizer.convert_ids_to_tokens(encoded_text.input_ids)
print(tokens)

['[CLS]', 'my', 'favorite', 'food', 'is', 'mas', '##ala', 'dos', '##a', '[SEP]']


In [41]:
tokenizer.convert_tokens_to_string(tokens)

'[CLS] my favorite food is masala dosa [SEP]'

In [42]:
tokenizer.vocab_size

30522

In [43]:
tokenizer.model_max_length

512

In [44]:
tokenizer.model_input_names

['input_ids', 'attention_mask']

## **TOKENIZATION ON DATASET**

In [45]:
emotion_dataset['train'][:2]

{'text': ["My favourite food is anything I didn't have to cook myself.",
  'Now if he does off himself, everyone will think hes having a laugh screwing with people instead of actually dead'],
 'labels': [[27], [27]],
 'id': ['eebbqej', 'ed00q6i']}

In [46]:
def tokenization(batch):
  return tokenizer(batch['text'], padding=True, truncation=True)

tokenization(emotion_dataset['train'][:2])

{'input_ids': [[101, 2026, 8837, 2833, 2003, 2505, 1045, 2134, 1005, 1056, 2031, 2000, 5660, 2870, 1012, 102, 0, 0, 0, 0, 0, 0, 0, 0], [101, 2085, 2065, 2002, 2515, 2125, 2370, 1010, 3071, 2097, 2228, 2002, 2015, 2383, 1037, 4756, 29082, 2007, 2111, 2612, 1997, 2941, 2757, 102]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}

        {'input_ids': [[101, 1045, 2134, 2102, 2514, 26608, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                      [101, 1045, 2064, 2175, 2013, 3110, 2061, 20625, 2000, 2061, 9636, 17772, 2074, 2013, 2108, 2105, 2619, 2040, 14977, 1998, 2003, 8300, 102]],
        'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
                            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}


In [47]:
emotions_encoded = emotion_dataset.map(tokenization, batched=True, batch_size=None)
emotions_encoded

Map:   0%|          | 0/43410 [00:00<?, ? examples/s]

Map:   0%|          | 0/5426 [00:00<?, ? examples/s]

Map:   0%|          | 0/5427 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['text', 'labels', 'id', 'input_ids', 'attention_mask'],
        num_rows: 43410
    })
    validation: Dataset({
        features: ['text', 'labels', 'id', 'input_ids', 'attention_mask'],
        num_rows: 5426
    })
    test: Dataset({
        features: ['text', 'labels', 'id', 'input_ids', 'attention_mask'],
        num_rows: 5427
    })
})

# **HuggingFace Models**

**BitsAndBytes:**

The bitsandbytes library is a Python library developed by OpenAI for efficient data serialization and deserialization.<br><br>

**Here's how it works:**

- Original List: [10, 20, 30, 40, 50]<br><br>
- Serialized Form: 10 20 30 40 50<br><br>
- In the serialized form, each number is represented as a string of characters separated by spaces. This representation may look shorter compared to the original list, but in terms of memory or storage size, it may not necessarily be shorter.<br><br>

- The primary benefit of serialization with bitsandbytes is not to make the data physically shorter but to convert it into a format that is easier to handle, transmit, or store, especially when dealing with complex data structures or when interoperating with other systems.<br><br>



**Accelerate:**<br><br>

- Suppose you're training a deep learning model to recognize images. This process involves lots of complex calculations.<br><br>
- With accelerate, you can harness the power of specialized hardware like GPUs (Graphics Processing Units) to speed up these calculations.<br><br>
- For example, if it takes 10 hours to train your model on a regular computer, accelerate might reduce it to just 2 hours on a GPU.<br><br>
- Similarly, when you're running your trained model to make predictions, accelerate helps make those predictions faster, which is crucial for real-time applications like video processing or natural language understanding.<br><br>
In summary, bitsandbytes helps with data compression and decompression, while accelerate boosts the performance of deep learning tasks by leveraging specialized hardware.














In [48]:
!pip install accelerate

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 309.4/309.4 kB 20.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.6/410.6 MB 850.0 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.1/14.1 MB 98.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.7/23.7 MB 34.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.6/823.6 kB 41.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 731.7/731.7 MB 1.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 121.6/121.6 MB 10.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.5/56.5 MB 15.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 124.2/124.2 MB 11.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 196.0/196.0 MB 1.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 176.2/176.2 MB 7.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 99.1/99.1 kB 9.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

# Mixtral-8x22B-v0.1

      "8x22" indicates that the model consists of 8 layers of 22 blocks each.
      "B" typically represents billion parameters, suggesting that the model has a large number of parameters.
      Now, let's discuss the concept of a Mixture of Experts (MoE) and how it's trained:

      Mixture of Experts (MoE):

      A Mixture of Experts is a neural network architecture that consists of multiple subnetworks called "experts" working in parallel, along with a gating mechanism that learns to dynamically select which expert to use for a given input. Each expert specializes in different aspects of the data, and the gating mechanism determines the relevance or importance of each expert for a particular input.

      In the context of language modeling, each expert might specialize in generating specific types of text, such as formal language, informal language, technical jargon, etc.

      Training Process:

      Expert Training: Initially, each expert in the MoE is trained independently on the data. This allows each expert to specialize in different aspects of the data.

      Gating Network Training: Simultaneously, the gating network is trained to learn how to dynamically combine the outputs of the experts based on the input data. The gating network's goal is to select the most relevant expert for each input.

      Joint Training: After training the experts and the gating network separately, the entire MoE model is fine-tuned jointly. During this joint training, the parameters of both the experts and the gating network are updated together to optimize the overall performance of the model.

      Example:

      Consider a language generation task where the input is a prompt, and the model is tasked with generating a coherent continuation of the prompt.

      Expert Networks: The MoE might consist of experts trained on different genres of text, such as news articles, fiction, scientific papers, etc.

      Gating Mechanism: The gating network learns to identify the genre or style of the input prompt and dynamically selects the expert best suited to generate text in that style.

      Training: During training, the MoE is fed a variety of prompts from different genres, and the parameters of both the experts and the gating network are adjusted to minimize the generation loss across all inputs.

      Overall, the Mixture of Experts architecture enables the model to leverage the strengths of multiple specialized subnetworks, resulting in improved performance and versatility across different types of data.








# Mixture of Experts Model Working with example

      Imagine you're planning a big event, like a party, and you need to organize various activities for different groups of guests. You decide to hire a team of experts, each specializing in a different aspect of event planning:

      Food Expert:
      This expert knows all about cooking and catering delicious meals.

      Music Expert:
      This expert specializes in creating playlists and setting up sound systems for entertainment.

      Decoration Expert:
      This expert is skilled at designing and decorating the venue to create a festive atmosphere.

      Entertainment Expert:
      This expert arranges games, performances, and other activities to keep guests entertained.

      Now, here's how the MoE architecture works:

      Expert Training:
      Each expert receives specialized training in their respective area. The food expert learns recipes, the music expert curates playlists, the decoration expert hones their design skills, and the entertainment expert plans fun activities.

      Gating Mechanism:
      You, as the event organizer, act as the gating mechanism. Based on the preferences and interests of your guests, you dynamically select which expert to consult for each aspect of the event. For example, if you have guests who love dancing, you might prioritize the music expert and choose lively playlists.

      Joint Training:
      As you plan the event, you interact with the experts and fine-tune the details. You might adjust the menu based on dietary preferences, tweak the playlist to suit different moods, and personalize the decorations to match the theme. This collaborative process optimizes the overall event experience.

      In summary, the MoE model architecture combines the expertise of specialized subnetworks (the experts) with a dynamic selection mechanism (the gating mechanism) to optimize performance across different tasks or domains. By leveraging the strengths of multiple experts, the MoE can adapt to diverse inputs and produce high-quality outputs tailored to specific needs.








In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer

model_id = "mistral-community/Mixtral-8x22B-v0.1"
tokenizer = AutoTokenizer.from_pretrained(model_id)

model = AutoModelForCausalLM.from_pretrained(model_id)

text = "Hello my name is"
inputs = tokenizer(text, return_tensors="pt")

outputs = model.generate(**inputs, max_new_tokens=4)
print(tokenizer.decode(outputs[0], skip_special_tokens=True))


model-00005-of-00059.safetensors:  93%|#########3| 4.48G/4.81G [00:00<?, ?B/s]

model-00006-of-00059.safetensors:   0%|          | 0.00/4.81G [00:00<?, ?B/s]

**AutoModelForCausalLM and AutoTokenizer**<br><br>
These are classes provided by the Hugging Face transformers library. They are used to simplify the process of loading and using pre-trained language models and tokenizers.<br><br>

**AutoModelForCausalLM:**

- This class is used to load pre-trained language models that are specifically designed for causal language modeling tasks.<br><br>
- Causal language modeling is a type of language modeling where the model predicts the next token in a sequence given the preceding tokens. It is commonly used for tasks like text generation and completion.<br><br>
- By using AutoModelForCausalLM, you can easily load a pre-trained model suitable for causal language modeling without worrying about the specific model architecture or configuration.<br><br>


**AutoTokenizer:**<br><br>

- This class is used to load pre-trained tokenizers that are compatible with specific language models.<br><br>
- Tokenizers are used to preprocess text data into tokens that can be fed into a language model.<br><br>
- The AutoTokenizer class automatically selects the appropriate tokenizer based on the specified model_id.

1.  **Table Question Answering from Tasks**




In [ ]:
from transformers import pipeline
import pandas as pd

# prepare table + question
data = {"Actors": ["Brad Pitt", "Leonardo Di Caprio", "George Clooney"], "Number of movies": ["87", "53", "69"]}
table = pd.DataFrame.from_dict(data)
question = "how many movies does Leonardo Di Caprio have?"

# pipeline model
# Note: you must to install torch-scatter first.
tqa = pipeline(task="table-question-answering", model="google/tapas-large-finetuned-wtq")

# result

print(tqa(table=table, query=question)['cells'][0])
#53


In [ ]:
print(tqa(table=table, query="bradd pitt movies")['cells'][0])

2.  **Document-Question-Answering from Tasks**

In [ ]:
from transformers import pipeline
from PIL import Image

pipe = pipeline("document-question-answering", model="naver-clova-ix/donut-base-finetuned-docvqa")

question = "What is the purchase amount?"
image = Image.open("/content/Screenshot 2024-04-13 021038.png")

pipe(image=image, question=question)




## **Evaluation Metrics**

# Bert Score

BERTScore is a metric used for evaluating the quality of text generated by language models. It measures how similar the generated text (predictions) is to the reference text.<br><br>

Here's a simpler breakdown:<br><br>

**What it measures:**<br><br>

- BERTScore calculates a score for each token in the generated text (predictions) by comparing it to each token in the reference text. It uses pre-trained language models like BERT to understand the context of the words.<br><br>

**How it works:**<br><br>

- BERTScore looks at each word in the generated text and finds the most similar word in the reference text using cosine similarity. It then calculates precision, recall, and F1 score based on these similarities.

**What it tells us:**<br><br>

- The precision score tells us how many words in the generated text match words in the reference text. The recall score tells us how many words in the reference text are captured by the generated text. The F1 score is a combination of precision and recall, providing an overall measure of similarity.<br><br>

**Usage:** <br><br>
- BERTScore can be used to compare different language generation models, evaluate the quality of machine translation, text summarization, and other natural language processing tasks.

In summary, BERTScore is a metric that measures the similarity between generated text and reference text, providing insight into the quality of language generation models. It's a useful tool for evaluating and comparing different text generation techniques.








In [ ]:
# Install the bert_score library
!pip install bert_score

# Import the necessary functions
from bert_score import score

# Prepare your predictions and references
predictions = ["hello world", "general kenobi"]
references = ["hello world", "general kenobi"]

# Compute BERTScore specifying the language (English)
precision, recall, f1 = score(predictions, references, lang="en")

# Print the results
print("Precision:", precision)
print("Recall:", recall)
print("F1 Score:", f1)


In [ ]:
!pip install evaluate

In [ ]:
from huggingface_hub import notebook_login

notebook_login()

In [ ]:
from evaluate import load
bertscore = load("bertscore")
predictions = ["hello there", "general kenobi"]
references = ["hello there", "general kenobi"]
results = bertscore.compute(predictions=predictions, references=references, lang="en")

# BLEU (Bilingual Evaluation Understudy):

**What it measures:**<br><br>

- BLEU is a metric used to evaluate the quality of machine-translated text by comparing it to human translations. It measures how closely the machine-generated text matches the human reference translations.<br><br>

**How it works:**<br><br>

      Segmentation:
      BLEU first divides the candidate translation (machine-generated text) and reference translations (human-generated text) into segments, usually sentences.

      N-gram matching:
      It then compares each segment of the candidate translation to the corresponding segments in the reference translations. BLEU considers the overlap of n-grams (contiguous sequences of n words) between the candidate and reference translations.

      Precision calculation:
      For each n-gram size, BLEU calculates precision, which measures the proportion of n-grams in the candidate translation that appear in the reference translations.

      Brevity penalty:
      BLEU also penalizes translations that are shorter than the reference translations to prevent shorter translations from receiving artificially high scores.

      BLEU score:
      Finally, BLEU combines the precision scores for different n-gram sizes using a geometric mean and applies the brevity penalty to compute the overall BLEU score

**Scoring range:**<br><br>
- BLEU scores range from 0 to 1, where a score closer to 1 indicates a better match between the machine-generated and human-translated texts. A score of 1 means the machine translation is identical to a human translation.<br><br>

**Usage:**<br><br>
- BLEU is commonly used for evaluating machine translation systems. It helps researchers and developers assess the quality of their translation models and make improvements.<br><br>

In summary, BLEU provides a quantitative measure of the quality of machine translations by comparing them to human translations. Higher BLEU scores indicate better quality translations. It's a widely-used and cost-effective metric for assessing translation performance.







In [ ]:
import nltk
from nltk.translate.bleu_score import sentence_bleu

# Define candidate and reference sentences
candidate = "hello world"
reference = ["hello world", "general kenobi"]

# Tokenize candidate and reference sentences
candidate_tokens = candidate.split()
reference_tokens = [ref.split() for ref in reference]

# Calculate BLEU score
bleu_score = sentence_bleu(reference_tokens, candidate_tokens)

# Print BLEU score
print("BLEU Score:", bleu_score)


# GLUE (General Language Understanding Evaluation)

- GLUE, or General Language Understanding Evaluation, is a test used to measure how well a machine understands human language. It includes different datasets to evaluate various aspects of language understanding.<br><br>

To use GLUE:<br><br>

**Load the relevant metric:**<br><br>

- Depending on which part of GLUE you're testing against, you load the corresponding metric.<br><br>

**Calculate the metric:**<br><br>

- Once loaded, provide your model's predictions and the reference answers. The metric then computes scores like accuracy or correlation.

**For example, if you're testing on a subset called MRPC, which checks accuracy and F1 score:**

      from evaluate import load

      # Load the GLUE metric for the MRPC subset
      glue_metric = load('glue', 'mrpc')

      # Prepare your predictions and references
      predictions = [0, 1]
      references = [0, 1]

      # Compute the GLUE score
      results = glue_metric.compute(predictions=predictions, references=references)

      # Print the results
      print(results)


**For STSB, which checks Pearson and Spearman correlations:**


        from evaluate import load

        # Load the GLUE metric for the STSB subset
        glue_metric = load('glue', 'stsb')

        # Prepare your predictions and references
        predictions = [-10., -11., -12., -13., -14., -15.]
        references = [0., 1., 2., 3., 4., 5.]

        # Compute the GLUE score
        results = glue_metric.compute(predictions=predictions, references=references)

        # Print the results
        print(results)


# PERPEXILITY

      Imagine you have a language model, like GPT-2, which has been trained on a large corpus of text data. Perplexity measures how well this model predicts a sequence of words.

      Let's say we have a simple sentence: "The cat sat on the mat." Now, the model tries to predict each word in this sentence based on the context of the previous words.

      Word Prediction: For each word in the sentence, the model predicts the probability distribution of the next word. For example:

      For "The", the model predicts the probability of "cat", "dog", "house", etc.
      For "cat", it predicts the probability of "sat", "jumped", "ate", etc.
      And so on for each word in the sentence.
      Perplexity Calculation: Perplexity calculates how surprised the model is by the actual words in the sequence. If the model is very confident in its predictions, the perplexity is low. If it's unsure and assigns similar probabilities to multiple words, the perplexity is high.

      Here's how it works:

      For each word in the sequence, the model assigns a probability to the actual word that appears next.
      Perplexity is the geometric mean of the inverse probabilities of the actual words. In simpler terms, it's like asking: "How surprised was the model by the actual sequence of words?"
      Let's say the model's predictions for our sentence are as follows:

      "The" -> "cat" (with probability 0.8)
      "cat" -> "sat" (with probability 0.7)
      "sat" -> "on" (with probability 0.6)
      "on" -> "the" (with probability 0.9)
      "the" -> "mat" (with probability 0.8)
      To calculate perplexity:

      Inverse the probabilities: 1/0.8, 1/0.7, 1/0.6, 1/0.9, 1/0.8.
      Calculate the geometric mean of these inverse probabilities.
      The result is the perplexity score.
      The lower the perplexity, the better the model's predictions match the actual sequence of words. In our example, if the perplexity is around 1, it means the model was very confident and not surprised by the actual sequence. If it's higher, it means the model was more uncertain about its predictions.






